Bech32 Library

jonathanknowles opened the pull request:
# Issue Number

Issue #238

jonathanknowles committed:
Add original implementation.

jonathanknowles committed:
Format imports and exports according to our coding standards.

jonathanknowles committed:
Format code according to our coding standards.

jonathanknowles committed:
Use <$> instead of `fmap .. $`.

jonathanknowles committed:
Import `Data.ByteString.Char8` as `B8` rather than `BSC`.

jonathanknowles committed:
Add original unit test source code.

jonathanknowles committed:
Format test code to meet our coding standards.

jonathanknowles committed:
Format imports and exports according to our coding standards.

jonathanknowles committed:
Change constraint from `Num` to `Integral`.

jonathanknowles committed:
Shorten length of `Word5` constructor from `UnsafeWord5` to `Word5`.

This makes code more concise, but doesn't affect safety, as this constructor is
not exported outside of the `Codec.Binary.Bech32` module.

jonathanknowles committed:
Move brackets to avoid unnecessary $.

jonathanknowles committed:
Import `Data.ByteString.Char8` as `B8` rather than `BSC`.

KtorZ committed:
add module description and author credits on top of the file

KtorZ committed:
replace usage of 'tasty' by 'hspec' in lib/bech32

KtorZ committed:
remove bitcoin-specific details and provide a slightly more general API

KtorZ committed:
add QuickCheck roundtrip for encode . decode

KtorZ reviewed (approved):
@jonathanknowles 

- I've replaced 'tasty' with 'hspec' in the tests for consistency across the project
- I've added some roundtrip fro the encoder and decoder
- I've removed the `segwit` stuff for this is bitcoin-specific. In our case, we'll be looking into https://github.com/input-output-hk/implementation-decisions/blob/master/text/0001-address.md (and the rust implementation as a reference). 

This leaves us with a basic encode / decode API for the bech32 format. I've slightly review the high-level module API to be easier to use (using `ByteString` as an input type instead of `[Word8]` and, removing type alias `HRP` in a favor of a newtype `HumanReadablePart` with a smart-constructor). 

This isn't perfect, but since I know you'll be working on some extra stuff for the error location, I'll leave it to you to iterate on top of this. Please don't spend too much time with this error detection though :pray:

jonathanknowles committed:
Add roundtrip tests for toBase256 and toBase32 functions.

jonathanknowles committed:
Add roundtrip tests for character set conversion functions.

jonathanknowles committed:
Add QuickCheck tests for `convertBits` function.

jonathanknowles committed:
Introduce `humanReadableCharset{Min,Max}Bound` constants.

Use these constants instead of repeating the same hard-coded values in
different places.

jonathanknowles committed:
Introduce constants `checksumLength` and `maxEncodedStringLength`.

Use these constants to avoid repeatedly hardcoding the same values
within functions.

jonathanknowles committed:
Improve documentation comment for function `convertBits`.

jonathanknowles committed:
Split module `Bech32` into public API and private `Internal` module.

KtorZ reviewed (approved):
:tada:
